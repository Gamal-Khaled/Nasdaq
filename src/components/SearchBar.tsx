import React, { useCallback, useRef, useState } from "react";
import { StyleSheet, TextInput } from "react-native";

import theme from "../theme/theme";
import { spacing } from "../theme/spacing";

interface Props {
  onSearch: (value: string) => void;
}

const SearchBarComponent = ({ onSearch }: Props) => {
  const [search, setSearch] = useState("");
  const debounceRef = useRef<NodeJS.Timeout>();

  const handleChangeText = useCallback((value: string) => {
    setSearch(value);
    if (debounceRef.current) {
      clearTimeout(debounceRef.current)
    }
    
    debounceRef.current = setTimeout(() => {
      onSearch(value);
    }, 300);
  }, [search]);


  const handleSubmit = useCallback(() => {
    onSearch(search);
  }, [search]);

  return (
    <TextInput
      style={styles.search}
      placeholder="Search for stocks"
      placeholderTextColor={theme.border}
      returnKeyType="search"
      onChangeText={handleChangeText}
      value={search}
      onSubmitEditing={handleSubmit}
    />
  );
};

const styles = StyleSheet.create({
  search: {
    borderWidth: 1,
    borderColor: theme.border,
    borderRadius: 30,
    marginHorizontal: spacing.base,
    marginVertical: spacing.base,
    paddingHorizontal: spacing.base,
    paddingVertical: spacing.small,
    color: theme.font,
  },
});

const SearchBar = React.memo(SearchBarComponent);
export default SearchBar;
