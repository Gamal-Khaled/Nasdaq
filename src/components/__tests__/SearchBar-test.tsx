import {
  act,
  cleanup,
  fireEvent,
  render,
  screen,
  userEvent,
} from "@testing-library/react-native";

import SearchBar from "../SearchBar";

afterEach(cleanup);

describe("Testing 'SearchBar' component", () => {
  it("Should render", () => {
    render(<SearchBar onSearch={jest.fn()} />);

    expect(screen.toJSON()).toMatchSnapshot();
  });
  test("Typing should change the input value and onSearch should be called only after timeout once", async () => {
    const mockSearchHandler = jest.fn();
    render(<SearchBar onSearch={mockSearchHandler} />);

    jest.useFakeTimers();
    const user = userEvent.setup();

    const input = screen.getByTestId("searchBar");
    await act(async () => {
      await user.type(input, "test");
    });

    expect(input).toHaveProp("value", "test");
    expect(mockSearchHandler).toHaveBeenCalledTimes(0);

    jest.runAllTimers();
    expect(mockSearchHandler).toHaveBeenCalledWith("test");
  });
  test("onSearch should be called when the user submits", async () => {
    const mockSearchHandler = jest.fn();
    render(<SearchBar onSearch={mockSearchHandler} />);

    jest.useFakeTimers();
    const user = userEvent.setup();

    const input = screen.getByTestId("searchBar");
    await act(async () => {
      await user.type(input, "test");
    });

    expect(input).toHaveProp("value", "test");
    expect(mockSearchHandler).toHaveBeenCalledTimes(0);

    fireEvent(input, "onSubmitEditing");
    expect(mockSearchHandler).toHaveBeenCalledWith("test");
  });
});
