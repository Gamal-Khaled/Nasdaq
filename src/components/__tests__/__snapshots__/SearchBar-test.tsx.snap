// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing 'SearchBar' component Should render 1`] = `
<TextInput
  onChangeText={[Function]}
  onSubmitEditing={[Function]}
  placeholder="Search for stocks"
  placeholderTextColor="#666980"
  returnKeyType="search"
  style={
    {
      "borderColor": "#666980",
      "borderRadius": 30,
      "borderWidth": 1,
      "color": "#fff",
      "marginHorizontal": 16,
      "marginVertical": 16,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
    }
  }
  testID="searchBar"
  value=""
/>
`;
